### ISSUE TITLE
Simulated keys are not shown with which-key

### ISSUE THREAD

**Issue by terlar** [noctuid/general.el#78]
Earlier when I tried, which-key displayed the mapped simulated key help when I pressed it. Currently nothing is shown by which-key when a simulated key is used.

For example (when using projectile):
`(general-define-key :keymaps '(normal visual) "SPC" (general-simulate-keys "C-c"))`

1. Press <kbd>SPC</kbd> (which-key shows nothing)
2. Press <kbd>p</kbd> (which-key shows just <kbd>p</kbd> was pressed and displays the bindings for <kbd>C-c</kbd><kbd>p</kbd>)

If you don't press another key which-key won't display any bindings for <kbd>C-c</kbd>.

**Comment by noctuid** [comment]
Sorry, I hadn't added documentation to the readme for this change yet. There are two ways of getting this behavior now with general's default method for `general--simulate-keys` (which is look up the key to try to match and run a command or match and set a keymap). 

You can set `which-key-show-transient-maps` to `t`. I'm not sure that this which-key feature works correctly in all cases. It's useful in other cases but annoying in some since it works for all transient maps.

The other (recommended) way is to set the no-lookup argument `(general-simulate-keys "C-c" nil nil t)`. I will change the default so that it won't use transient maps. The reason for the current default is because simulating keys in a way that triggers which-key breaks evil repeat/macro recording. General now fixes the issue during macro/repeat playback, so using transient maps by default isn't necessary anymore.

I'll close this after adding documentation to the readme.

**Comment by terlar** [comment]
Thank you that solution is working. It was vaguely hinted in the documentation, but I couldn't figure out how to get that specific behavior. So a clarification there would be greatly appreciated. Once again thank you for your work.

**Comment by yatesco** [comment]
I ran into this.

Previously, this:
``` emacs
(general-nvmap :prefix "SPC"
...
 "p" (general-simulate-keys
       "C-c p" t
       "Projectile"
       general-SPC-simulates-C-c-p)
...)
```
Showed "Projectile" if I pressed  `SPC`. After a refresh it now fails trying to interpret "Projectile" as a `symbol`.

The following works, but it sets the `which-key` text to "general-SPC-simulates-C-c-p":
``` emacs
"p"  (general-simulate-keys
          "C-c p" t
           nil
           "Projectile")
```

Any hints (other than 'yeah, learn elisp! ;-)')

**Comment by terlar** [comment]
If I understand correctly that function takes `KEYS &optional STATE KEYMAP NO-LOOKUP DOCSTRING NAME`.

So I guess you could try to use:
`(general-simulate-keys "C-c p" nil nil t "Projectile" general-SPC-simulates-C-c-p)`

**Comment by noctuid** [comment]
That's correct. All general functions/macros have docstrings, so if you're in doubt, you can always `C-h f funtion-name`.

**Comment by yatesco** [comment]
Thanks both, but that still shows "general-SPC-simulates-C-c-p" in `which-key`. I can live with that for sure.

**Comment by noctuid** [comment]
If I remember correctly, `C-c p` is bound to `projectile-command-map`. This would be preferable: `"p" '(:keymap projectile-command-map :wk "projectile")`. In either case, you currently can't set the which key replacement directly from `general-simulate-keys`.

**Comment by yatesco** [comment]
That's perfect @noctuid - thanks!

**Comment by noctuid** [comment]
I've tried to avoid any serious breaking changes, but in this case, since there have already been breaing changes to `general-simulate-keys`, I'm consider changing it further. If I make changes, what I'd probably do is use keyword arguments everything after `KEYMAP` for consistency with the other helpers (the key/predicate dispatch macros) and to make it easier to add extra arguments in the future (e.g. I may add `WHICH-KEY` to specify a which-key replacement like with `general-key-dispatch`). Any feedback on this would be appreciated.

**Comment by yatesco** [comment]
Makes sense to me. 

If you are worried about breaking changes then you could also use a new function and mark the old one as deprecated (assuming emacs’ lisp has such a notion).

My vote for the new name is ‘press-this-instead’ :-).

**Comment by noctuid** [comment]
I've fixed this issue in the `simulate` branch. I've added tests but would appreciate if someone else could also try it before I merge. All optional arguments are replaced with keyword arguments. The new macro is called `general-simulate-key` and the old `general-simulate-keys` is marked as obsolete. I'm also planning to add a `:which-key` keyword argument and another similar macro that I'm going to recommend using instead of `general-simulate-key` whenever possible.
